src/Components/Server/src/Endpoints/CircuitMatcherPolicy.cs

```diff
@@ -0,0 +1,232 @@
+                    return;
+                }
+
+                await endpoint.RequestDelegate(context);
+            }
+
+            var routeEndpointBuilder = new RouteEndpointBuilder(
+                ProtectedEndpoint,
+                routeEndpoint.RoutePattern,
+                routeEndpoint.Order);
+
+            // Preserve the metadata
+            foreach (var metadata in endpoint.Metadata)
+            {
+                routeEndpointBuilder.Metadata.Add(metadata);
+            }
+
+            return routeEndpointBuilder.Build();
+        }
+
+        private bool ValidateCircuitId(HttpContext context)
+        {
+            if (!context.Request.Query.TryGetValue(QueryStringParameterKey, out var circuitId) || circuitId.Count != 1)
+            {
+                // There is no circuit id on the query string, so we don't authenticate anything.
+                return false;
+            }
+
+            // Signalr requires clients to send the header X-Requested-With or fails.
+            // This means that it will force cross-origin requests to go through CORS
+            // which means we will see a preflight request, in which case we fail.
+            if (HttpMethods.IsOptions(context.Request.Method) &&
+                context.Request.Headers.ContainsKey(HeaderNames.AccessControlRequestMethod))
+            {
+                return false;
+            }
+
+            var key = $"{CookiePrefix}{GetCircuitIdPrefix(circuitId)}";
+
+            if (context.Request.Cookies.TryGetValue(key, out var cookie))
+            {
+                if (CircuitIdFactory.ValidateCircuitId(circuitId, cookie, out var parsedId))
+                {
+                    CleanupStaleCookies(context);
+
+                    var identity = new ClaimsIdentity();
+                    identity.AddClaim(new Claim(IdClaimType, parsedId.RequestToken));
+                    context.User.AddIdentity(identity);
+
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        // Blazor requires stickyness as the circuit is held in memory.
+        // That means that on scale-out scenarios there must be a session cookie and all circuits for a given
+        // session will end up in the same server.
+        // With this information, we can simply check the cookies the client sends us and discard any cookie for
+        // which we can't find a circuit, either connected or disconnected, as when the user closes the browser
+        // all sessions (and whatever affinity cookie was set) will go away.
+        // Circuits for which there isn't an already created circuit will contain an additional KeepAlive cookie
+        // that will be used as a filter to avoid deleting the circuit id cookie. The KeepAlive cookie for a given
+        // circuit will go away automatically.
+        private void CleanupStaleCookies(HttpContext context)
```

Comment:
I didn't spot any E2E (or unit) tests covering this. Did I miss this? Will some be added?

Reply:
It does work (I verified manually) but I chose not to add one because it takes until the circuit goes away to take effect, so I didn't want to add a test that waits for 4 minutes.

In retrospective I think I might be able to add an endpoint that pings the server and tells it to dispose a circuit out of band, and that would allow for a test this in an E2E fashion